starlark/int.go

```diff
@@ -142,75 +138,161 @@ var (
-func (x Int) Rsh(y uint) Int { return Int{new(big.Int).Rsh(x.bigint, y)} }
+	if i.bigint != nil {
+		// TODO(adonovan): opt: handle common values without allocation.
+		f, _ := new(big.Float).SetInt(i.bigint).Float64()
+		return Float(f)
+	}
+	return Float(float64(i.val))
+}
+
+func (x Int) Sign() int {
+	switch {
+	case x.bigint != nil:
+		return x.bigint.Sign()
+	case x.val > 0:
+		return 1
+	case x.val < 0:
+		return -1
+	}
+	return 0
+}
+
+func (x Int) Add(y Int) Int {
+	if x.bigint != nil || y.bigint != nil {
+		return Int{bigint: new(big.Int).Add(x.big(), y.big())}.maybeSmall()
+	}
+	return Int{val: x.val + y.val}.maybeBig()
+}
+func (x Int) Sub(y Int) Int {
+	if x.bigint != nil || y.bigint != nil {
+		return Int{bigint: new(big.Int).Sub(x.big(), y.big())}.maybeSmall()
+	}
+	return Int{val: x.val - y.val}.maybeBig()
+}
+func (x Int) Mul(y Int) Int {
+	if x.bigint != nil || y.bigint != nil {
+		return Int{bigint: new(big.Int).Mul(x.big(), y.big())}.maybeSmall()
+	}
+	return Int{val: x.val * y.val}.maybeBig()
+}
+func (x Int) Or(y Int) Int {
+	if x.bigint != nil || y.bigint != nil {
+		return Int{bigint: new(big.Int).Or(x.big(), y.big())}
+	}
+	return Int{val: x.val | y.val}
+}
+func (x Int) And(y Int) Int {
+	if x.bigint != nil || y.bigint != nil {
+		return Int{bigint: new(big.Int).And(x.big(), y.big())}
+	}
+	return Int{val: x.val & y.val}
+}
+func (x Int) Xor(y Int) Int {
+	if x.bigint != nil || y.bigint != nil {
+		return Int{bigint: new(big.Int).Xor(x.big(), y.big())}
+	}
+	return Int{val: x.val ^ y.val}
+}
+func (x Int) Not() Int {
+	if x.bigint != nil {
+		return Int{bigint: new(big.Int).Not(x.bigint)}
+	}
+	return Int{val: ^x.val}
+}
+func (x Int) Lsh(y uint) Int { return Int{bigint: new(big.Int).Lsh(x.big(), y)}.maybeSmall() }
+func (x Int) Rsh(y uint) Int { return Int{bigint: new(big.Int).Rsh(x.big(), y)}.maybeSmall() }
 
 // Precondition: y is nonzero.
 func (x Int) Div(y Int) Int {
+	xb, yb := x.big(), y.big()
 	// http://python-history.blogspot.com/2010/08/why-pythons-integer-division-floors.html
 	var quo, rem big.Int
-	quo.QuoRem(x.bigint, y.bigint, &rem)
-	if (x.bigint.Sign() < 0) != (y.bigint.Sign() < 0) && rem.Sign() != 0 {
-		quo.Sub(&quo, one.bigint)
+	quo.QuoRem(xb, yb, &rem)
+	if (xb.Sign() < 0) != (yb.Sign() < 0) && rem.Sign() != 0 {
+		quo.Sub(&quo, oneBig)
 	}
-	return Int{&quo}
+	return Int{bigint: &quo}.maybeSmall()
 }
 
 // Precondition: y is nonzero.
 func (x Int) Mod(y Int) Int {
+	xb, yb := x.big(), y.big()
+
 	var quo, rem big.Int
-	quo.QuoRem(x.bigint, y.bigint, &rem)
-	if (x.bigint.Sign() < 0) != (y.bigint.Sign() < 0) && rem.Sign() != 0 {
-		rem.Add(&rem, y.bigint)
+	quo.QuoRem(xb, yb, &rem)
+	if (xb.Sign() < 0) != (yb.Sign() < 0) && rem.Sign() != 0 {
+		rem.Add(&rem, yb)
+	}
+	return Int{bigint: &rem}.maybeSmall()
+}
+
+func (i Int) rational() *big.Rat {
+	if i.bigint != nil {
+		return new(big.Rat).SetInt(i.bigint)
+	}
+	return new(big.Rat).SetInt64(i.val)
+}
+
+func (i Int) maybeBig() Int {
```

Comment:
I think maybeBig wants to be a function from int64 to Int, rather than a function that takes an Int whose representation invariant is violated and returns a sound Int. Similarly maybeSmall should be a function from *big.Int to Int. In other words, make this the implementation of MakeInt64 and call that function instead of maybeBig. Similarly, add a new public function MakeBigInt, and call that instead of maybeSmall.

Reply:
And since we're adding a constructor from big.Int, we should have an accessor big values too. Let's rename the toBig method (again) to BigInt for symmetry.

We should also document at the MakeBigInt and BigInt methods that the client most not modify the provided/returned big.Int variable.

Reply:
Added the MakeBigInt, BigInt methods and move toBig to newBig.